README
; Copyright (c) 1993-2008 by Richard Kelsey and Jonathan Rees. See file COPYING.

Please report bugs to scheme-48-bugs@s48.org, and include
the version number in your message.

Installation instructions in file INSTALL.

A user's guide is in file doc/user-guide.txt.

Recent changes are listed in file doc/news.txt.

Known bugs and things to do in the future are listed in doc/todo.txt.

Send mail to scheme-48-request@s48.org to be put on a
mailing list for announcements, discussion, bug reports, and bug
fixes.

-----

The Scheme 48 root directory is organized as follows (not all files are
listed here):

    README		this file
    WINDOWS.txt		notes on the Windows port
    INSTALL		installation instructions
    COPYING		copyright notice
    COPYING.rtf		copyright notice in RTF format
    configure		configuration script
    Makefile.in		input to configure
    doc/		some documentation
      scheme48.man	a Unix-style manual page
      manual.ps		manual in Postscript
      manual.pdf        manual in PDF
      html/		manual in html (the root is html/index.html)
      src/		manual sources
      todo.txt		list of improvements we hope to make someday
      news.txt		list of improvements we have already made
      io.txt		how the I/O system works
    scheme/		scheme source files
      packages.scm	meta-module definitions
      interfaces.scm	system interface definitions
      more-interfaces.scm system interface definitions
      *-packages.scm	module definitions
      bcomp/		the byte-code compiler
      vm/		virtual machine sources (written in Pre-Scheme)
      rts/		run-time system sources
      link/		static linker
      env/		development environment modules (debugger, etc.)
      big/		useful Scheme libraries and extensions ("Big Scheme")
      cml/		Concurrent ML libraries
      alt/		portable implementations of some Scheme 48 features
      opt/		optional code optimizer for the byte-code compiler
      prescheme/	code for running the VM using Scheme 48
      debug/		debugging utilities, tests, etc.
      misc/		very miscellaneous things (e.g. AMB operator)
      posix/		interface to POSIX system calls
      srfi/		SRFI implementations
      sort/		sorting libraries
    ps-compiler/	Pre-Scheme -> C compiler
    c/			c source files
      sysdep.h.in	input to configure
      scheme48vm.c	most of the VM (generated by Pre-Scheme compiler)
      scheme48vm.h	extern declarations for scheme48vm.c
      scheme48vm-prelude.h internal declarations for scheme48vm.c
      scheme48heap.c    storage management (generated by Pre-Scheme compiler)
      scheme48heap.h	extern declarations for scheme48heap.c
      scheme48read-image.c    reading a heap from a file
      scheme48write-image.c   writing a heap to a file
      scheme48image.h	extern declarations for scheme48...-image.c
      scheme48write-barrier.h WRITE_BARRIER(...) macro
      main.c		entry point for the VM
      prescheme.h	part of the VM
      bignum.c		large integers
      bignum.h          declarations for large integers
      bignumint.h	internal declarations for large integers
      external.c	support for calling C and being called from C
      extension.c       default definition of vm_extension()
      scheme48.h	C declarations and macros for Scheme 48 data structures
      scheme48.h.in	template for scheme48.h
      old-scheme48.h	old version, included for compatibility
      c-mods.h		minor additions to the C language
      event.h		header file for OS interface
      io.h		ditto
      fd-io.h		ditto
      srfi-27.c         C-side support for SRFI 27 (random bits)
      asm-glue.c	support code for (upcoming) native code
      unix/             Unix-specific source files
      unix.h		header file for unix/* files
      posix/		C half of interface to POSIX system calls
      fake/		C files for insufficiently POSIX-compliant systems
    build/		code for building the system
      filenames.make	included by Makefile, generated automatically
      filenames.scm	code for generating filenames.make
      initial.image	an image file containing a minimal Scheme system
      initial.debug	debugging database for same
      initial.scm	script for creating initial.image
      build-usual-image	script for creating scheme48.image
      load-linker.exec  script for loading the external linker
      build-external-modules script for creating external-module initializer
      minor-version-number   current version number
      generate-c-header.bat Windows batch file for generating c/scheme48.h
      build-usual-image.bat Windows batch file for creating scheme48.image
      build-initial-image.bat Windows batch file for creating build/initial.image
      i-know-what-i-am-doing.bat Windows batch file for compiling VM to C
      scheme48.wxs      WiX script for generating scheme48.msi
      build/{UnicodeData.txt,PropList.txt,SpecialCasing.txt,CaseFolding.txt,CompositionExclusions.txt}
                        Unicode data files
    emacs/		Emacs support


doc/news.txt
-*- Mode: org; -*-

Recent changes to Scheme 48.

* 2012-12-31 (version 1.9)
** Additions
- A new C FFI as added, thanks to Crestani, Harald Glab-Plhak.  The
  old one is still there but will eventually be phased out.
  See Crestani's paper "Foreign-Function Interfaces for Garbage-Collected Programming Languages"
  http://www-pu.informatik.uni-tuebingen.de/users/crestani/publications/2008-sws-ffi.pdf
- A new networking code layer was added with full support for IPv6,
  UDP etc. (undocumented as of yet)
- Records now support single inheritance (accessible through the
  r6rs-records packages)
- `letrec*' was added to the `scheme' structure
- A `r5rs' structure was added.
- A new condition system based on R6RS's system has replaced the old system.
- Some R6RS libraries were added (with help from Robert Ransom).
- A statistical profiler was added (Marcel Turino, Manuel Dietrich) -
  check the documentation for details.
- TLC tables (`eq?' tables) were added by Marcus Crestani and David Frese.
- The module system warns on cycles and redefinitions
- When the VM is compiled with GNU C, it uses direct threading for
  instruction dispatch, which speeds up the VM noticeable (done by
  Timo Harter)
- The reader for a structure is configurable (again).
- SRFI 19 now works on Windows
- `list-spine{,-cycle-safe}[*%]' sequence macros were added to `reduce'
  structure by Robert Ransom
- The Windows build can be done via a Boo script (contributed by
  Robert Ransom)
- The unquote and unquote-splicing were generalized to several
  operands, in line with R6RS.
- Commands ,show-known-packages, ,show-interface, and
  ,show-default-package were added to support SDT (Marcus Crestani,
  Sebastian Rheinecker).
- `placeholder-value' now accepts a `deadlock?' argument that says
  whether blocking on the placeholder contributes to deadlock.
- Most source files now carry author information.
- A document =doc/deriving.txt= was added that clarifies how works
  derived from Scheme 48 should be labelled.
** Changes
- The BIBOP GC was heavily debugged and is now considered stable - it
  is enabled by default.
- The system builds on Windows Visual Studio Express 2010 instead of
  Visual Studio 2005.
- The BIBOP GC code was simplified significantly.
- The `syntax-rules' implementation was rewritten by Richard Kelsey
- Some structures were added to the POSIX subsystem 
  (`posix-errnos', `posix-syslog')
- Various procedures were added to the POSIX structures by Roderic
  Morris to support scsh.
- The autoconf code was reorganized by Ivan Shmakov.
- The REPL now prints values without quotes and auto-capitalization of
  record-type names.
- The Makefile now builds the documentation upon `make install'.
  (This requires a working LaTeX installation and tex2page.)  A target
  `install-no-doc' is available that does not.
- The license information was clarified.
- The external-events API was changed to fix a design bug; it now
  correctly accomodates "fire-once" applications such as
  wait-`for-child-process' (with help from Robert Ransom, Roderic
  Morris).
** Bug fixes
- Many.

* 01/20/08 (version 1.8)
** Additions
- An experimental framework for asynchronous events from external code
  was added.  It's now used for asynchronous host-name lookups on
  systems with pthreads.
- Support for 64-bit architectures was added (thanks to Taylor
  Campbell, Marcus Crestani, and David Frese)
- The installation now includes an (experimental) "scheme48-config"
  script that reports C compiler and linker and associated flags that
  may be used for compiling dynamic externals.
** Changes
- The layout of the installed system now conforms to FHS.
- The old dynamic-loading code was removed; correspondingly, the -o
  flag for scheme48vm was removed.
- The filename translations apply to dynamic externals as well.
- On Unix, the central event loop uses poll() instead of select() on
  systems that have it (thanks to Marcus Crestani)
- The external code for the POSIX libraries and SRFI 27 was moved to
  dynamically-loaded externals.  (Cygwin support for this thanks to
  Harald Glab-Plhak)
** Bug fixes
- A bug in the division simplifier of the PreScheme compiler was
  fixed.
- The PreScheme compiler can now emit code for unsigned longs.
- The implementations of SRFI 19 and SRFI 42 now include the latest
  fixes from the reference implementations.
- A race condition in the implementation of queues was fixed.

* 09/06/07 (version 1.7)
** Additions
- A new, experimental garbage collector written by David Frese has
  been added.  It can be enabled by specifying "--enable-gc=bibop" on
  the configure command line.  (It is disabled by default on Unix, but
  enabled on Windows.)
- DESTDIR support has been added to the Makefile (thanks to Markus
  Weissmann, Ivan Shmakov, Will Job von Foo)
** Bug fixes
- Ctrl-C now works more reliably on Windows.
- An I/O bug on Windows where an EOF on read would cause a crash has
  been fixed.  (reported by Christoph Bauer)
- A bug that would cause spurious deadlocks with the interactive REPL
  was fixed.
- A bug where the last exception handler would get de-installed has
  been fixed (reported by Emilio Lopes)
- A bug in BYTES->STRING-N that would cause it to break on incomplete
  incodings was fixed (reported by Emilio Lopes)
- CURRENT-OS-STRING-TEXT-CODEC is now exported from OS-STRINGS as
  documented (reported by Emilio Lopes)
- A bug where random text would be printed when a port was closed by
  GC was fixed.  (reported by Will Clinger)
- A bug in the UDP sockets was fixed.

* 03/19/07 (version 1.6)
** Additions
- An autogen.sh script for generating the various generated files from
  source was added.
- Infrastructure for creating, running and composing test suites was
  added.  As a result, "make check" now runs more tests than before.
- It is now possible for adventurous users to run the linker in PLT
  Scheme.
** Changes
- "make image" now assumes a Unicode-capable Scheme 48.
** Bug fixes

c/init.c
/* Part of Scheme 48 1.9.  See file COPYING for notices and license.
   */

#include <stdlib.h>
#include <stdio.h>
#include "scheme48vm.h"
#include "scheme48heap.h"
#include "scheme48image.h"
#include "ffi.h"

extern long s48_get_file_size(unsigned char *);

#if !defined(DEFAULT_HEAP_SIZE)
/* 4 megacells = 16 megabytes on 32-bit systems */
#define DEFAULT_HEAP_SIZE 4000000L
#endif

#if !defined(DEFAULT_STACK_SIZE)
/* 2500 cells = 10000 bytes */
#define DEFAULT_STACK_SIZE 2500L
#endif

#if defined(STATIC_AREAS) && defined(S48_GC_BIBOP)
#error "The BIBOP GC doesn't support the STATIC_AREAS feature yet."
#endif

#if defined(STATIC_AREAS)
#define DEFAULT_IMAGE_NAME NULL
#else

/* DEFAULT_IMAGE_NAME should be defined using the -D switch to cc. */
#if !defined(DEFAULT_IMAGE_NAME)
#define DEFAULT_IMAGE_NAME "scheme48.image"
#endif

#endif /* STATIC_AREAS */

extern void	s48_sysdep_init(void);
extern void	s48_initialize_external_modules(void);

long
s48_initialize(int *argcp, char ***argv)
{
  char *image_name = DEFAULT_IMAGE_NAME;
  long heap_size = DEFAULT_HEAP_SIZE;    /* in numbers of cells */
  long stack_size = DEFAULT_STACK_SIZE;  /* in numbers of cells */
  int errors = 0;
  char *stack;

#if defined(STATIC_AREAS)
  extern long static_entry;
  extern long static_symbol_table;
  extern long static_imported_binding_table, static_exported_binding_table;
  extern long p_count, *p_areas[], p_sizes[];
  extern long i_count, *i_areas[], i_sizes[];
#endif

  int argc = *argcp;
  int vm_argc = 0;
  char *me = *(*argv);		/* Save program name. */

  {
    /* initialize floating-point printer */
    extern void s48_free_init(void);

    s48_free_init();
  }

  (*argv)++; argc--;		/* Skip program name. */

  for (; argc > 0; argc--, (*argv)++)
    if ((*argv)[0][0] == '-')
      switch ((*argv)[0][1]) {
      case 'h':
	argc--; (*argv)++;
	if (argc == 0) { errors++; break; }
	heap_size = atol(*(*argv));
	if (heap_size < 0) errors++;  /* 0 means now no limit */
	break;
      case 's':
	argc--; (*argv)++;
	if (argc == 0) { errors++; break; }
	stack_size = atoi(*(*argv));
	if (stack_size <= 0) errors++;
	break;
      case 'i':
	argc--; (*argv)++;
	if (argc == 0) { errors++; break; }
	image_name = *(*argv);
	break;
      case 'a':
	argc--;
	vm_argc = argc;    /* remaining args are passed to the VM */
	argc = 0;
	break;
      case 'I':
#ifdef S48_GC_BIBOP
	heap_size = 0;  /* unlimited heap size (BIBOP GC only) */
#endif
	/* code for -i */
	argc--; (*argv)++;
	if (argc == 0) { errors++; break; }
	image_name = *(*argv);
	/* code for -a */
	argc--;
	vm_argc = argc;    /* remaining args are passed to the VM */
	argc = 0;
	break;
      default:
	fprintf(stderr, "Invalid argument: %s\n", *(*argv));
	errors++;
      }
    else
      if ((*argv)[0][0] != '\0') {
	fprintf(stderr, "Invalid argument: %s\n", *(*argv));
	errors++; }
  if (errors != 0) {
    fprintf(stderr,
"Usage: %s [options] [-a arguments]\n\
Options: -h <heap-size>    %s heap size in words (default %ld).%s\n\
	 -s <stack-size>   Stack buffer size in words.\n\
         -i <file>         Load image from file (default \"%s\")\n",
	    me,
#if S48_GC_BIBOP
	    "Maximum",
	    (long)DEFAULT_HEAP_SIZE,
"\n                           A heap size of 0 means the heap can grow\n\
                           unboundedly. This is dangerous because it can\n\
                           cause your system to run out of memory.",
#else
	    "Total",
	    (long)DEFAULT_HEAP_SIZE,
	    "",
#endif
	    DEFAULT_IMAGE_NAME
	    );
    return 1;
  }


  /* Disable GC to read the image and initialize the VM-stack */ 
  s48_forbid_gcB();

  s48_sysdep_init();
  s48_heap_init();
  s48_init();

  if (image_name == NULL) {
#if defined(STATIC_AREAS)
    s48_register_static_areas(p_count, p_areas, p_sizes,
			      i_count, i_areas, i_sizes);
    s48_set_image_valuesB(static_entry,
			  static_symbol_table,
			  static_imported_binding_table,
			  static_exported_binding_table);
    if (s48_initialize_heap(heap_size) == NULL) {
      fprintf(stderr, "system is out of memory\n");
      return 1; }
#else
    fprintf(stderr, "No image file.\n");
    return 1;
#endif
  } else if (s48_read_image(image_name, heap_size) == -1) {
    fprintf(stderr, "Image file \"%s\" is unusable.\n", image_name);
    return 1; }
  
  stack = (char *) malloc(stack_size * sizeof(long));
    
  if (!stack) {
    fprintf(stderr, "system is out of memory\n");
    return 1; }

  s48_initialize_ffi();

  s48_initialize_vm(stack, stack_size);

  s48_initialize_external_modules();

  /* Heap und stack are ok. Enable the GC. */
  s48_allow_gcB();

  *argcp = vm_argc;

  return 0;
}
